### DISCUSSION TITLE
Understanding Feature Importance/Engineering

### DISCUSSION THREAD

**Discussion by ebagjones** [timeseriesAI/tsai#270] [Q&A]
Are there any tools or recommended approaches to determining feature importance with tsai? I'm particularly interested in a method that would work with (mini)ROCKET but I'm all ears if anyone knows of a good approach with the other algorithms. Thanks and Regards.

**Comment by vrodriguezf** [comment]
Do you mean the importance of the features generated by ROCKET or the  the importance with respect to the original time series variables? In case it's the first, it's simple (any tabular approach with good interpretability such as xgb can do it), but I'm not sure if it is useful at all since the features are random kernels.

If you are interested in interpretability I wouldn't go with ROCKET. Instead, you can try NN architectures like XCM in which you can create activation maps in both the time and the variable dimensions, which allows for a sort of heatmap that could give you the feature importance.

**Reply by ebagjones** [reply]
@vrodriguezf thanks for the reply. I'm interested in knowing the importance of the original time-series signals. I would assume that the time series signals may be ranked differently importance or signal contribution vs a stationary snapshot of a single point like I'd get with a RF or xgboost classification. Any guidance on this would be much appreciated.

**Reply by vrodriguezf** [reply]
Have a look at [`get_attribution_map`](https://github.com/timeseriesAI/tsai/blob/main/nbs/100c_models.explainability.ipynb), there you can pass data to a trained NN, and get a "heatmap" of how each point in the sequence contributes to the activations and gradients of one specific layer. Normally you would use it for the last module of the backbone of the NN, in Plus models that have backbone and head.

**Reply by ebagjones** [reply]
@vrodriguezf  this is very cool, thanks for pointing me in the right direction!

**Comment by oguiza** [comment]
Hi @jefferyanderson and @vrodriguezf,
I've just released some new code to allow you to plot feature importance using permutation feature importance (issue #286). This shows the drop in the model's performance when a single feature is shuffled. 
It's important to note that it doesn't show the importance of the feature along the time axis. 
To use this new functionality all you need to do is this: 
```bash
X, y, splits = get_UCR_data('LSST', split_data=False)
tfms = [None, TSClassification()]
batch_tfms = TSStandardize(by_sample=True)
dls = get_ts_dls(X, y, splits=splits, tfms=tfms, batch_tfms=batch_tfms)
learn = ts_learner(dls, InceptionTimePlus, metrics=accuracy, cbs=[ShowGraph()])
learn.fit_one_cycle(10, 1e-2)
learn.feature_importance()
```
If you try it please, let me know if it works well.

**Reply by vrodriguezf** [reply]
this is super useful and elegant :)

**Reply by oguiza** [reply]
Thanks, Victor! It'll be great if it also works well!! 😃

**Reply by ebagjones** [reply]
@oguiza wow, thank you very much. I'll very busy with work this week but I'll definitely try it out by this weekend. Thanks in advance, this looks very exciting.

**Reply by ebagjones** [reply]
@oguiza sorry for what is almost certainly a dumb question, but I pulled the latest (pip install -Uqq git+https://github.com/timeseriesAI/tsai.git) and I'm getting thsi when I run your example:

```
----> 7 learn.feature_importance()

~/.local/lib/python3.8/site-packages/tsai/learner.py in feature_importance(self, feature_names, key_metric_idx, show_chart, save_df_path, random_state)
    323     # Adapted from https://www.kaggle.com/cdeotte/lstm-feature-importance by Chris Deotte (Kaggle GrandMaster)
    324     if feature_names is None:
--> 325         feature_names = [f"var_{i}" for i in range(X.shape[1])]
    326 
    327     results = []

NameError: name 'X' is not defined
```
is installing via pip this way incorrect?

**Reply by oguiza** [reply]
Thanks for raising this! This is a bug in the code. X should be X_valid. I just made a patch release to GitHub and it should be fixed now (issue #291). Please, try it again. 
pip install -Uqq git+https://github.com/timeseriesAI/tsai.git  is fine to upgrade tsai from GitHub.

**Reply by strakehyr** [reply]
I used the feature_importance and it works great! I was wondering if (since this is a time-series focused library) there are any plans on enhancing the feature_importance method with a way to output the feature importance per timestep (importance of feature var_2 at t_24 for instance).
Also wanted to know if there are any plans on integrating the library with SHAP/Lime.

**Reply by scottcha** [reply]
I was trying this functionality out today.  For my data, which is larger than memory and I've been using the recommended zarr storage from the larger than memory tutorial I get OOM errors when using this.  It seems to be related to OOM in the get_X_preds.  I tried digging in to it but was unable to determine exactly why I'm hitting these.

**Reply by oguiza** [reply]
> are any plans on enhancing the feature_importance method with a way to output the feature importance per timestep?

At timeseriesAI we are constantly reviewing papers looking for better time series models and methods. We actually included a model that integrates this approach: XCM. If you are interested you can review the [paper](https://hal.inria.fr/hal-03469487/document) that's been published earlier this week. You can find more information on tsai's implementation in the [documentation](https://timeseriesai.github.io/tsai/models.XCM.html). 

> Also wanted to know if there are any plans on integrating the library with SHAP/Lime.

No, there are no immediate plans to integrate any of those methods. You can actually read about some comparisons between SHAP/ Lima and XCM in the paper I shared before. 

Having said that, we are open to new ideas that improve explainability.

**Reply by oguiza** [reply]
Thanks for raising this @scottcha.
Could you please create an issue with this (including the full traceback) so that we can take a closer look at it?
Also, are you interested in running the method on all samples (which may be slow for large datasets), or would it be acceptable to pass a smaller, in-memory subset randomly chosen to determine feature importance?

**Reply by scottcha** [reply]
@oguiza thanks, I just submitted the issue here: https://github.com/timeseriesAI/tsai/issues/310.  Its acceptable to pass a smaller in-memory subset in my case.
Thanks!

**Reply by ebagjones** [reply]
@oguiza this is an awesome feature, thanks again for adding it.

**Reply by ebagjones** [reply]
@oguiza the new feature importance looks awesome! Thank you.

**Reply by oguiza** [reply]
Thanks a lot for your feedback @jefferyanderson ! It's much appreciated 🤩

**Reply by strakehyr** [reply]
Just wanted to give my kudos to the team for the step_importance() feature that I recently found out. Really nice.

**Reply by oguiza** [reply]
Thanks, @strakehyr! 🙂
